packages/driver/src/cy/commands/actions/attachFile.ts

```diff
@@ -6,23 +6,111 @@ import $errUtils from '../../../cypress/error_utils'
+      }
+    }
+  } catch (e) {
+    // We're in Firefox, this is fine.
+  }
+
+  return item
+}
+
+const createDataTransfer = (files: Cypress.FileReferenceObject[]): DataTransfer => {
   const dataTransfer = new DataTransfer()
 
   files.forEach(({ contents, fileName, lastModified, mimeType }) => {
-    const file = new File([contents], fileName || '', { lastModified, type: mimeType })
+    dataTransfer.items.add(new File([contents], fileName || '', {
+      lastModified,
+      type: mimeType,
+    }))
+  })
+
+  const oldItems = dataTransfer.items
 
-    dataTransfer.items.add(file)
+  // dataTransfer.items is a getter, and it - and the items read from its
+  // returned DataTransferItemList - cannot be assigned to. DataTransferItemLists
+  // also cannot be constructed, so we have to use an array instead.
+  Object.defineProperty(dataTransfer, 'items', {
+    get () {
+      return _.map(oldItems, tryMockWebkit)
+    },
   })
 
-  return dataTransfer.files
+  // When a real user drags file(s) over an element, dataTransfer.types always contains `['Files']`.
+  // In Firefox however, the 'types' array is not properly populated when we construct our dataTransfer.
+  if (files.length !== 0 && dataTransfer.types.length === 0) {
+    Object.defineProperty(dataTransfer, 'types', {
+      get () {
+        // This is the value observed from real user events in Firefox 90.
+        return ['application/x-moz-file', 'Files']
+      },
+    })
+  }
+
+  return dataTransfer
 }
 
 interface InternalAttachFileOptions extends Cypress.AttachFileOptions {
   _log: any
   $el: JQuery
 }
 
+const ACTIONS = {
+  input: (element, dataTransfer, coords, state) => {
+    (element as HTMLInputElement).files = dataTransfer.files
+    const inputEventOptions = addEventCoords({
+      bubbles: true,
+      composed: true,
+    }, coords)
+
+    const changeEventOptions = addEventCoords({
+      bubbles: true,
+    }, coords)
+
+    dispatch(element, state('window'), 'input', inputEventOptions)
+    dispatch(element, state('window'), 'change', changeEventOptions)
+  },
+  'drag-n-drop': (element, dataTransfer, coords, state) => {
+    const dragEventOptions = addEventCoords({
+      bubbles: true,
+      composed: true,
+      cancelable: true,
+      dataTransfer,
+    }, coords)
+
+    dispatch(element, state('window'), 'drag', dragEventOptions)
+    dispatch(element, state('window'), 'dragenter', dragEventOptions)
+    dispatch(element, state('window'), 'dragover', dragEventOptions)
+    dispatch(element, state('window'), 'drop', dragEventOptions)
+
+    // If a user drops file over an <input type="file"> element, browsers attach the file
+    // to it, exactly as if they'd selected one from a file list.
+    // If a user drops multiple files over an input without the "multiple" attribute,
+    // the browser does nothing, and we want to mimic this behavior.
+    if ($dom.isInputType(element, 'file') && (dataTransfer.files.length === 1 || element.multiple)) {
```

Comment:
New to this type of functionality, should this conditional have an else? or if this element is not an input type, nothing is needed?

Reply:
Yeah, if the target is not an input element, then we don't need to do anything more.

Basically, if you drop a file on a div, you get ['drag', 'dragenter', 'dragover', 'drop']. If you drop one file on a file input (or any number of files on a `multiple` file input), then you *also* get ['input', 'change'] events firing.